datafusion/sqllogictest/test_files/map.slt

```diff
@@ -100,3 +100,93 @@ physical_plan
 
 statement ok
 drop table table_with_map;
+
+query ?
+SELECT MAKE_MAP('POST', 41, 'HEAD', 33, 'PATCH', 30, 'OPTION', 29, 'GET', 27, 'PUT', 25, 'DELETE', 24) AS method_count;
+----
+{POST: 41, HEAD: 33, PATCH: 30, OPTION: 29, GET: 27, PUT: 25, DELETE: 24}
+
+query I
+SELECT MAKE_MAP('POST', 41, 'HEAD', 33)['POST'];
+----
+41
+
+query ?
+SELECT MAKE_MAP('POST', 41, 'HEAD', 33, 'PATCH', null);
+----
+{POST: 41, HEAD: 33, PATCH: }
+
+query ?
+SELECT MAKE_MAP('POST', null, 'HEAD', 33, 'PATCH', null);
+----
+{POST: , HEAD: 33, PATCH: }
+
+query ?
+SELECT MAKE_MAP(1, null, 2, 33, 3, null);
+----
+{1: , 2: 33, 3: }
+
+query ?
+SELECT MAKE_MAP([1,2], ['a', 'b'], [3,4], ['b']);
+----
+{[1, 2]: [a, b], [3, 4]: [b]}
+
+query error DataFusion error: Execution error: map requires all values to have the same type Int64, got Utf8 instead at position 1
+SELECT MAKE_MAP('POST', 41, 'HEAD', 'ab', 'PATCH', 30);
+
+query error DataFusion error: Execution error: make_map key cannot be null at position 2
+SELECT MAKE_MAP('POST', 41, 'HEAD', 33, null, 30);
+
+query error DataFusion error: Execution error: make_map requires all keys to have the same type Utf8, got Int64 instead at position 1
+SELECT MAKE_MAP('POST', 41, 123, 33,'PATCH', 30);
+
+query error
+SELECT MAKE_MAP()
+
+query error DataFusion error: Execution error: make_map requires an even number of arguments, got 3 instead
+SELECT MAKE_MAP('POST', 41, 'HEAD');
+
+query ?
+SELECT MAP(['POST', 'HEAD', 'PATCH'], [41, 33, 30]);
+----
+{POST: 41, HEAD: 33, PATCH: 30}
+
+query ?
+SELECT MAP(['POST', 'HEAD', 'PATCH'], [41, 33, null]);
+----
+{POST: 41, HEAD: 33, PATCH: }
+
+query ?
+SELECT MAP([[1,2], [3,4]], ['a', 'b']);
+----
+{[1, 2]: a, [3, 4]: b}
+
+query error
+SELECT MAP()
+
+query error DataFusion error: Execution error: map requires an even number of arguments, got 1 instead
+SELECT MAP(['POST', 'HEAD'])
+
+query error DataFusion error: Execution error: Expected list, large_list or fixed_size_list, got Null
+SELECT MAP(null, [41, 33, 30]);
+
+query error DataFusion error: Execution error: map requires key and value lists to have the same length
+SELECT MAP(['POST', 'HEAD', 'PATCH'], [41, 33]);
+
+query error DataFusion error: Execution error: map key cannot be null
+SELECT MAP(['POST', 'HEAD', null], [41, 33, 30]);
+
+query ?
+SELECT MAP(make_array('POST', 'HEAD', 'PATCH'), make_array(41, 33, 30));
+----
+{POST: 41, HEAD: 33, PATCH: 30}
+
+query ?
+SELECT MAP(arrow_cast(make_array('POST', 'HEAD', 'PATCH'), 'FixedSizeList(3, Utf8)'), arrow_cast(make_array(41, 33, 30), 'FixedSizeList(3, Int64)'));
+----
+{POST: 41, HEAD: 33, PATCH: 30}
+
+query ?
+SELECT MAP(arrow_cast(make_array('POST', 'HEAD', 'PATCH'), 'LargeList(Utf8)'), arrow_cast(make_array(41, 33, 30), 'LargeList(Int64)'));
+----
+{POST: 41, HEAD: 33, PATCH: 30}
```

Comment:
I think if possible we should also add tests for array values (not just scalars)

I wrote up some examples, like this:

```
# test that maps can be created from arrays
statement ok
create table t as values
('a', 1, ['k1', 'k2'], [10.0, 20.0]),
('b', 2, ['k3'], [30.0]),
('d', 4, ['k5', 'k6'], [50.0, 60.0]);

query error
select make_map(column1, column2) from t;
----
DataFusion error: Internal error: UDF returned a different number of rows than expected. Expected: 3, Got: 1.
This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker


query error
select map(column3, column4) from t;
----
DataFusion error: Execution error: Expected scalar, got ListArray
[
  StringArray
[
  "k1",
  "k2",
],
  StringArray
[
  "k3",
],
  StringArray
[
  "k5",
  "k6",
],
]

```

Reply:
Oops. I think I misunderstood how the `ColumnarValue` works. I expected it to be inputted per row, but it inputs all the rows at once.
```
[datafusion/functions/src/core/map.rs:59:5] &key = [
    StringArray
    [
      "a",
      "b",
      "d",
    ],
]
[datafusion/functions/src/core/map.rs:60:5] &value = [
    PrimitiveArray<Int64>
    [
      1,
      2,
      4,
    ],
]
``` 
I'll fix them.